swarm/src/lib.rs

```diff
@@ -415,18 +415,81 @@ where
 
         let peer_id = dial_opts.get_peer_id();
         let condition = dial_opts.peer_condition();
+        let dial_conditions = dial_opts.dial_conditions();
         let connection_id = dial_opts.connection_id();
 
-        let should_dial = match (condition, peer_id) {
-            (PeerCondition::Always, _) => true,
-            (PeerCondition::Disconnected, None) => true,
-            (PeerCondition::NotDialing, None) => true,
-            (PeerCondition::Disconnected, Some(peer_id)) => !self.pool.is_connected(peer_id),
-            (PeerCondition::NotDialing, Some(peer_id)) => !self.pool.is_dialing(peer_id),
+        enum DialOrPeerCondition {
+            PeerCondition(PeerCondition),
+            DialCondition(DialConditions),
+        }
```

Comment:
Can we avoid this enum if we do some early returns or something?

I'd like to optimize this code for _deletion_. i.e. once we decide to remove the deprecated functionality, I'd like to just be able to delete a block of code and not have to refactor this.

Does that make sense?